[Improvement](Iceberg) Optimize LocationPath.of performance for Iceberg table queries

suxiaogang223 opened the pull request:
### What problem does this PR solve?

### Problem
When querying Iceberg tables with large number of data files, `LocationPath.of` method consumes significant CPU time. The main bottlenecks are:
1. Repeated regex parsing in `S3URI.create()` for each file path
2. Multiple `String.split()` calls for scheme extraction
3. Repeated `StorageProperties` lookup from map for each file

### Solution
This PR introduces several optimizations to reduce CPU overhead:

#### 1. Optimize scheme parsing in `LocationPath.parseScheme`
- Replace `String.split("://")` with `indexOf` + `substring` to avoid array allocation

#### 2. Add fast path for S3-compatible schemes in `S3PropertyUtils.validateAndNormalizeUri`
- For simple S3-compatible URIs like `oss://bucket/key`, `s3a://bucket/key`, use direct string replacement instead of full S3URI regex parsing
- Only fall back to full S3URI parsing for complex HTTP URLs

#### 3. Add path prefix caching in `IcebergScanNode`
- Cache `StorageProperties`, schema, and path prefix mapping on first file
- For subsequent files with the same prefix, directly transform paths using string replacement
- New `LocationPath.ofDirect()` method to create LocationPath without any parsing

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

suxiaogang223 committed:
optmize location

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/source/IcebergScanNode.java:
```diff
@@ -363,9 +375,83 @@ private CloseableIterable<FileScanTask> planFileScanTask(TableScan scan) {
+        if (locationPathCacheInitialized) {
+            return;
+        }
+        synchronized (this) {
```
No need synchronized, because scan node can only be visited one thread

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/source/IcebergScanNode.java:
```diff
@@ -363,9 +375,83 @@ private CloseableIterable<FileScanTask> planFileScanTask(TableScan scan) {
+            } catch (Exception e) {
+                // If caching fails, we'll fall back to the full method each time
+                LOG.warn("Failed to initialize LocationPath cache, will use full parsing", e);
+                locationPathCacheInitialized = true;
```
Do we need to reset all cached values?

morningman reviewed (commented):
Add UT for new methods

suxiaogang223 committed:
fix

suxiaogang223 committed:
Enhance location validation and caching in LocationPath and S3PropertyUtils

- Added authority checks in LocationPath to throw exceptions for invalid locations.
- Improved caching mechanism in IcebergScanNode for location path creation.
- Updated S3PropertyUtils to handle empty and invalid paths during scheme conversion.
- Added unit tests for new functionality and edge cases in LocationPathTest and S3ConnectorPropertiesUtilsTest.

suxiaogang223 committed:
Remove invalid test cases from S3ConnectorPropertiesUtilsTest to streamline unit tests for URI validation.

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  60.53% (46/76)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/59217_873a9f03ff4fa04f1113ca11df44fc1410574fec_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/59217_873a9f03ff4fa04f1113ca11df44fc1410574fec_merge_fe/report/index.html)

kaka11chen reviewed (approved):
LGTM
